services/data-info/src/data_info/routes/avus.clj

```diff
@@ -12,26 +13,29 @@
 
     (GET* "/avus" [:as {uri :uri}]
       :query [{:keys [user]} StandardUserQueryParams]
-      :return AVUGetResult
+      :return MetadataListing
+      :middlewares [wrap-metadata-base-url]
       :summary "List AVUs (administrative)"
-      :description (str "List iRODS AVUs associated with a data item. Include administrative/system AVUs."
-(get-error-code-block "ERR_NOT_A_USER, ERR_NOT_READABLE"))
+      :description
+          (str "List all AVUs associated with a data item. Include administrative/system iRODS AVUs."
+(get-error-code-block "ERR_NOT_A_USER, ERR_NOT_READABLE")
+(get-endpoint-delegate-block "metadata" "GET /avus/{target-type}/{target-id}"))
       (svc/trap uri meta/admin-metadata-get data-id))
 
-    (POST* "/avus" [:as {uri :uri}]
+    (PUT* "/avus" [:as {uri :uri}]
       :query [{:keys [user]} StandardUserQueryParams]
-      :body [{:keys [irods-avus]} (describe AVUListing "An list of AVUs to add")]
+      :body [{:keys [irods-avus]} (describe AVUListing "A list of AVUs to add")]
       :return AVUChangeResult
       :summary "Add AVUs (administrative)"
-      :description (str "Associate AVUs with a data item. Allow adding any AVU."
+      :description (str "Associate iRODS AVUs with a data item. Allow adding any AVU."
 (get-error-code-block "ERR_NOT_A_USER, ERR_NOT_READABLE, ERR_DOES_NOT_EXIST, ERR_NOT_WRITEABLE, ERR_NOT_AUTHORIZED"))
       (svc/trap uri meta/admin-metadata-add data-id irods-avus))
```

Comment:
The highlight I did was intended to show basically what you're saying, that PUT shouldn't be used when it's not a complete representation of the resource at that particular URI (in this case, the list of AVUs), so it needs to be set and not just add ("modified version of the one residing on the origin server" meaning "a complete representation of the resource", since this is REST). PATCH would... probably work for add/update, but it's not really well-defined in our code or generally how PATCH should be implemented so most of the time POST makes more sense (since POST is sort of the catch-all sending some data to a data handling process). Plus, we don't support deletes with this sort of endpoint so it's not a complete patch.

I'd argue that the set is still idempotent with respect to meaningful values in this case. Since the new IDs various AVUs get are random, the preservation of a particular random value doesn't really affect the semantics of idempotence -- as long as after N calls that value is still a new, random value, it's still indistinguishable from running it only once (the fact that it was a different random value in the middle isn't meaningfully different). Arguably we either have useless IDs (we don't actually care much about preserving them as far as I know; other than figuring out how to address endpoints concerning single AVUs I don't think we need them, and we could just change that) or we should change the implementation to preserve IDs even when they aren't provided if there's an appropriate non-surrogate identifier (possibly all three values of the AVU, in this case -- for iRODS AVUs that's good enough but I'm not completely sure of the semantics of our metadata). Either change would make that endpoint properly idempotent.

Basically, I think the idempotence issue is an artifact of the fact we're using sort of poorly defined surrogate identifiers, and the other guidance should be referred to instead.